File: 157. Read N Characters Given Read4.cpp
// Forward declaration of the read4 API.
int read4(char *buf);

class Solution {
public:
    /**
     * @param buf Destination buffer
     * @param n   Maximum number of characters to read
     * @return    The number of characters read
     */
    int read(char *buf, int n) {
        char local[4];
        int read = 0, cur = 0, i = 0;
        while ((read = read4((char*)&local))) {
            read = cur + read <= n ? cur + read : n;
            for (i = 0; cur < read; ++cur, ++i) {
                buf[cur] = local[i];
            }
        }
        return cur;
    }
};

File: java/157. Read N Characters Given Read4.java
/* The read4 API is defined in the parent class Reader4.
      int read4(char[] buf); */
// O(n)
public class Solution extends Reader4 {
    /**
     * @param buf Destination buffer
     * @param n   Maximum number of characters to read
     * @return    The number of characters read
     */
    public int read(char[] buf, int n) {
        int i = 0, retNum;
        char[] readBuf = new char[4];
        while (i != n) {
            retNum = read4(readBuf);
            int count = Math.min(retNum, n - i);
            for (int j = 0; j < count; ++j) {
                buf[i++] = readBuf[j];
            }
            if (retNum < 4) {
                break;
            }
        }
        return i;
    }
}
